perf/smallbank_workload/src/smallbank_tranformer.rs

```diff
@@ -80,3 +143,30 @@ impl<'a> SBPayloadTransformer<'a> {
         Ok(txn)
     }
 }
+
+struct DependencyTracker<T>
+where
+    T: Eq + Hash,
+{
+    signature_by_id: HashMap<T, String>,
+}
+
+impl<T> DependencyTracker<T>
+where
+    T: Eq + Hash,
+{
+    pub fn new() -> DependencyTracker<T> {
+        DependencyTracker {
+            signature_by_id: HashMap::new(),
+        }
+    }
+
+    pub fn get_dependency_for_id(&self, id: &T) -> Option<&String> {
+        self.signature_by_id.get(id)
+    }
+
+    pub fn add_signature_for_id(&mut self, id: T, signature: String) {
```

Comment:
Call this `add_dependency()`

Reply:
So, as I see it,
```
# 1
add_signature(..., id: T, signature: String)
get_dependency(..., id: T)

#2
add_signature(..., id: T, signature: String)
get_signature(..., id: T)

#3
add_dependency(..., id: T, signature: String)
get_dependency(..., id: T)
```
I think I like #2 the most as it introduces the least amount of indirection. Although the case for #1 is that when it is `add_signature` the signature is actually the signature for the payload that introduces `id` and `get_dependency` gets the dependency for any payload that uses `id`. I'm less clear on the advantages of #3.

Reply:
3 makes the most sense to me. I think it is preferable over 1, because there is strong conceptual link between the two methods. I think it is preferable over 2, because you are tracking dependencies, not signatures. In our architecture, dependencies are specified by giving the transaction signature, but we could easily be using something else (say for example, if we used a hash of the payload as the transaction id instead of the signature).

Can you elaborate on your definition of indirection? I define indirection to mean roughly "having code in a separate function/method/class". Typically, this is for the purposes of maintaining a consistent level of abstraction in different portions of the code and to allow reuse of code. With that definition, I don't see how any of the above options have different levels of indirection, but I am interested to hear what your definition is.

Reply:
I guess I don't have a great definition of indirection, so I won't try to make one up. What I was trying to say was that using `_dependency` adds a layer of meaning that might not always be correct, whereas `get_signature` and `add_signature` is the plain meaning of those methods.

I am fine having pretty bedrock architectural decisions of Sawtooth affect the nature of a piece  of code that creates Sawtooth Transacions.